--- /usr/local/share/perl/5.30.0/Bio/Graphics/Glyph/xyplot.pm	2021-01-21 20:51:13.000000000 +1100
+++ /usr/local/share/perl/5.30.0/Bio/Graphics/Glyph/xyplot.pm.fix	2021-01-27 17:25:29.049725473 +1100
@@ -311,8 +311,8 @@
   my @parts    = $self->parts;
   my $lw       = $self->linewidth;
   # Make the boxes transparent
-  my $positive = $self->pos_color + 1073741824;
-  my $negative = $self->neg_color + 1073741824;
+  my $positive = $self->pos_color;#AP + 1073741824;
+  my $negative = $self->neg_color;#AP + 1073741824;
   my $height   = $self->height;
 
   my $midpoint = $self->midpoint ? $self->score2position($self->midpoint) 
